### PR TITLE
Add Dashboard using Plotly Dash.

### PR DESCRIPTION
**Pull request by Hephaestus12**

### CODE REVIEW

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    return estimates['total_count'], estimates['austria_count'], estimates['belgium_count'], estimates['germany_count'], estimates['italy_count'], estimates['netherlands_count']
+
+# Nothing -> main graph
+@app.callback(
+    Output("main_graph", "figure"),
+    [Input("year_selector", "value")],
+)
+def make_main_figure(years):
+
+    # print(df['austria'])
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]: These comments should be deleted. You could also keep them by logging their output somewhere if you need them for debug purposes.

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+)
+def make_histogram(model, countries, years):
+
+    cases = {}
+    for year in years:
+        # print(year)
+        for country in countries:
+            # print(country)
+            for index, row in df[country].iterrows():
+                # print("Row is: ")
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]: Same here

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+# Create callbacks
+app.clientside_callback(
+    ClientsideFunction(namespace="clientside", function_name="resize"),
+    Output("output-clientside", "children"),
+    [Input("count_graph", "figure")],
+)
+
+
+# Radio -> multi
+@app.callback(Output("country_names", "value"), [Input("country_selector", "value")])
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]: It would be nice to add a couple of lines that explain what do these methods do.

File: gsoc_application_projects/2020/influenza/dash/assets/resizing_script.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+if (!window.dash_clientside) {
+  window.dash_clientside = {};
+}
+window.dash_clientside.clientside = {
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/assets/resizing_script.js:4]: If you took this code from somewhere, you should cite the source. Not super important since there is no proper copyright on it (i think) but it is just to acknowledge the person which wrote it.

File: gsoc_application_projects/2020/influenza/dash/random_forest/model.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  Linear Ridge Regression.
+"""
+import logging
+import numpy as np
+import shogun as sg
+import config
+from random_forest import util
+
+
+class Model:
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/random_forest/model.py:14]: I reckon that you initially wrote a different piece of code to instantiate a model (`model/src/model.py`), why did you write another one here? Couldn't you use the previous one?

File: gsoc_application_projects/2020/influenza/dash/util.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+import pageviewapi
+import pandas as pd
+from pageviewapi.client import ZeroOrDataNotLoadedException, \
+    ThrottlingException
+
+import config
+from random_forest import model
+
+
+class DataGateway:
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/util.py:23]: Again here, why did you copy code around and you didn't use the one you wrote during the previous iteration? Did you need to add new functionalities or was it not performing the way we wanted?

**Review by geektoni — CHANGES_REQUESTED** [review]
I've added some commits of general nitpicking :) I am just wondering why there are so many classes copied directly from the previous code for the web application. Is it possible to reuse them or do you plan to delete the old ones instead?

File: gsoc_application_projects/2020/influenza/dash/util.py (hunk 6 above)
Comment (Hephaestus12) [reply gsoc_application_projects/2020/influenza/dash/util.py:23]: Here, I've made the `web` and `dash` as standalone applications, when I make the docker container, I'll just be copying the dash folder to that and it'll work. That's why this file had to be transferred here. Also, the old and new web tools are basically two entirely different python flask web applications.
About the `model` directly, it is simply the notebooks and scripts which were used to process the data and develop the models.
Basically, the three directories are all independent.

Comment (Hephaestus12) [reply gsoc_application_projects/2020/influenza/dash/util.py:23]: > Here, I've made the `web` and `dash` as standalone applications, when I make the docker container, I'll just be copying the dash folder to that and it'll work. That's why this file had to be transferred here. Also, the old and new web tools are basically two entirely different python flask web applications.
> About the `model` directly, it is simply the notebooks and scripts which were used to process the data and develop the models.
> Basically, the three directories are all independent.

Is this okay? Or should I make any changes?

**Comment by Hephaestus12** [comment]
Commented code removed.
Now this saves calculated estimates as well :)

**Comment by Hephaestus12** [comment]
All that's left now is to upload it to dockerhub.

**Comment by Hephaestus12** [comment]
https://hub.docker.com/r/tejsukhatme/influenza_estimator <-------done :)

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 7]
```diff
@@ -266,33 +251,21 @@ def update_text(years):
...
-    # a = 0
-    # for key in estimates:
-        # a += estimates[key]
-        # estimates[key] = int(estimates[key])
-    # a = int(a)
-
     return estimates['total_count'], estimates['austria_count'], estimates['belgium_count'], estimates['germany_count'], estimates['italy_count'], estimates['netherlands_count']
 
 # Nothing -> main graph
+# Controls What is diaplyed on the main map
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]:
```suggestion
# Controls What is displayed on the main map
```

File: gsoc_application_projects/2020/influenza/dash/util.py (hunk 6 above)
Comment (geektoni) [reply gsoc_application_projects/2020/influenza/dash/util.py:23]: mmmh I am not sure if I understand how the things are structured now. Why do we have two separate applications and not just the one made with dash? or is it just one? Is there any reason why you kept also the old version of the interface? I am asking because, generally, you should avoid code duplication. Here, it seems that you have basically copy-pasted files around instead of reusing the ones you already wrote.

**Comment by geektoni** [comment]
I was able to pull the new docker image and it looks very nice :) I also tried to play with the various options in the control panel but they do not seem to work yet. Secondly, I've seen that there are also other graphs which are still empty though. Could you elaborate a list of things which are missing and you still need to implement to have all the new interface up and running?

**Comment by Hephaestus12** [comment]
> I was able to pull the new docker image and it looks very nice :) I also tried to play with the various options in the control panel but they do not seem to work yet. Secondly, I've seen that there are also other graphs which are still empty though. Could you elaborate a list of things which are missing and you still need to implement to have all the new interface up and running?

I don't understand? Everything works when I run it, Can you send me a screenshot of your screen?

**Comment by geektoni** [comment]
> > I was able to pull the new docker image and it looks very nice :) I also tried to play with the various options in the control panel but they do not seem to work yet. Secondly, I've seen that there are also other graphs which are still empty though. Could you elaborate a list of things which are missing and you still need to implement to have all the new interface up and running?
> 
> I don't understand? Everything works when I run it, Can you send me a screenshot of your screen?

I just checked again and the other graphs indeed work (not for all the possible combinations though). Would it be possible to set already some default values in the control panel such to have some data displayed form the beginning?

**Comment by Hephaestus12** [comment]
Yes, all possible combinations don't work, primarily because of the years.
For example, in the dataset, I only have access to the pageviews from one country for the year 2008. As a result if you choose just 2008 and some other country, it shows zero cases. I'll try to reduce the ambiguity in this.

File: gsoc_application_projects/2020/influenza/dash/assets/resizing_script.js (hunk 4 above)
Comment (Hephaestus12) [reply gsoc_application_projects/2020/influenza/dash/assets/resizing_script.js:4]: Done.

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 8]
```diff
@@ -254,7 +272,7 @@ def update_text(years):
     return estimates['total_count'], estimates['austria_count'], estimates['belgium_count'], estimates['germany_count'], estimates['italy_count'], estimates['netherlands_count']
 
 # Nothing -> main graph
-# Controls What is diaplyed on the main map
+# Controls What is displyed on the main map
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]:
```suggestion
# Controls What is displayed on the main map
```

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 9]
```diff
@@ -419,6 +458,45 @@ def make_line(model, countries, years):
     return fig
 
 
+
+## REST API
+server = app.server
+
+# @server.route('/')
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]: remove

**Comment by geektoni** [comment]
Apart from some nitpicking now it seems to be right :) are these latest changes already on docker-hub?

**Comment by Hephaestus12** [comment]
> Apart from some nitpicking now it seems to be right :) are these latest changes already on docker-hub?

Thanks, and yes, it is published to docker-hub.

File: gsoc_application_projects/2020/influenza/dash/Dockerfile [hunk 10]
```diff
@@ -29,9 +29,6 @@ COPY . /
...
 RUN pip --no-cache-dir install -r requirements.txt
 
-ENV FLASK_APP=influenza_estimator
-ENV FLASK_ENV=development
+EXPOSE 8050
 
-EXPOSE 5000
-
-CMD ["flask", "run", "-h", "0.0.0.0", "-p", "5000"]
+CMD ["python", "app.py", "-h", "0.0.0.0", "-p", "5000"]
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/Dockerfile]: What is the purpose of these flags (`-h` and `-p`)? I think they are not used by `app.py`.

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+    week = str(year) + '-' + str(week)
+    for country in COUNTRIES:
+        incidence = df[country].at[week, 'incidence']
+        ans[country] = incidence
+    return json.dumps(ans)
+
+
+# Main
+if __name__ == "__main__":
+    app.run_server(host='0.0.0.0', debug=False, port=8050)
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]: Ahh I see, you hardcoded the various information here. Wouldn't it be possible to let the user specify the address/port so that he/she won't need to touch directly the source code?

File: gsoc_application_projects/2020/influenza/dash/app.py [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+    ],
+)
+def enable_years(countries):
+
+    year_disable = [True for i in range(14)]
+    for country in countries:
+        for year in country_years[country]:
+            year_disable[year-2007] = False
+        
+    print(year_disable)
```
Comment (geektoni) [inline gsoc_application_projects/2020/influenza/dash/app.py]: You forgot this `print` statement :)

File: gsoc_application_projects/2020/influenza/dash/app.py (hunk 11 above)
Comment (Hephaestus12) [reply gsoc_application_projects/2020/influenza/dash/app.py]: I can remove it from here, but it'll have to be specified in some particular config file. I don't believe it's possible to set these at runtime.

Comment (geektoni) [reply gsoc_application_projects/2020/influenza/dash/app.py]: mmh I don't think it is mandatory. You can provide this information as command-line arguments. See https://docs.python.org/3/library/argparse.html#module-argparse

Comment (Hephaestus12) [reply gsoc_application_projects/2020/influenza/dash/app.py]: Ohh right, this is so cool. I'll implement this. :D
